firedantic/tests/tests_async/test_model.py

```diff
@@ -534,4 +536,116 @@ async def test_save_with_exclude_unset(configure_db) -> None:
+async def test_create_in_transaction(configure_db) -> None:
+    """Test creating a model in a transaction. Test case from README."""
+
+    @async_transactional  # type: ignore
+    async def create_in_transaction(transaction, email) -> Profile:
+        """Creates a Profile in a transaction"""
+        try:
+            await Profile.find_one({"email": email}, transaction=transaction)
+            raise ValueError(f"Profile already exists with email: {email}")
+        except ModelNotFoundError:
+            p = Profile(email=email)
+            await p.save(transaction=transaction)
+            return p
+
+    transaction = CONFIGURATIONS["db"].transaction()
+    p = await create_in_transaction(transaction, "test@example.com")
+    assert isinstance(p, Profile)
+    assert p.id
+
+    transaction2 = CONFIGURATIONS["db"].transaction()
+    with pytest.raises(ValueError) as excinfo:
+        await create_in_transaction(transaction2, "test@example.com")
+    assert str(excinfo.value) == "Profile already exists with email: test@example.com"
+
+
+@pytest.mark.asyncio
+async def test_delete_in_transaction(configure_db) -> None:
+    """Test deleting a model in a transaction."""
+
+    @async_transactional  # type: ignore
+    async def delete_in_transaction(
+        transaction: AsyncTransaction, profile_id: str
+    ) -> dict:
+        """Deletes a Profile in a transaction."""
+        try:
+            p = await Profile.get_by_id(profile_id, transaction=transaction)
+            await p.delete(transaction=transaction)
+            return {"id": profile_id}
+        except ModelNotFoundError:
+            raise ValueError(f"Profile not found: {profile_id}")
+
+    p = Profile(name="Foo")
+    await p.save()
+    assert p.id
+
+    transaction = CONFIGURATIONS["db"].transaction()
+    result = await delete_in_transaction(transaction, p.id)
+    assert isinstance(result, dict)
+    assert result == {"id": p.id}
+
+    transaction2 = CONFIGURATIONS["db"].transaction()
+    with pytest.raises(ValueError) as excinfo:
+        await delete_in_transaction(transaction2, p.id)
+    assert str(excinfo.value) == f"Profile not found: {p.id}"
+
+
+@pytest.mark.asyncio
+async def test_update_model_in_transaction(configure_db) -> None:
+    """Test updating a model in a transaction."""
+
+    @async_transactional  # type: ignore
+    async def update_in_transaction(
+        transaction: AsyncTransaction, profile_id: str, name: str
+    ) -> Profile:
+        """Updates a Profile in a transaction."""
+        p = Profile(id=profile_id)
+        await p.reload(transaction=transaction)
+        p.name = name
+        await p.save(transaction=transaction)
+        return p
+
+    p = Profile(name="Foo")
+    await p.save()
+    assert p.id
+
+    transaction = CONFIGURATIONS["db"].transaction()
+    result = await update_in_transaction(transaction, p.id, "Bar")
+    assert isinstance(result, Profile)
+    assert result.id == p.id
+    assert result.name == "Bar"
```

Comment:
Would maybe not be a bad idea to also reload `p` and that way verify the copy in the database was actually also updated. Now this merely tests that the assignment (in the python code alone) to the model (that was returned) was successful.

Would also be nice if the names inside the inner function (`p`) would not shadow the ones from the outer scope, as it makes it a bit hard to keep track of which `p` is which. Alternatively you could maybe pass in the profile and then in the transactional function just reload it instead of passing in the id of it and making a new object of it.

Reply:
I've set the result to be reloaded after it's retrieved from the transaction but before the name assertion is checked.  This way we know it's actually gotten saved and retrieved from nthe database successfully.

I've updated the inner functions so they use profile instead of p. This way the names do not overlap.